nemo/collections/asr/parts/mixins/ipl_mixin.py

```diff
@@ -0,0 +1,679 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import os
+import random
+import tempfile
+from typing import List, Union
+
+import editdistance
+import torch
+from omegaconf import ListConfig, open_dict
+from tqdm.auto import tqdm
+
+from nemo.collections.asr.data.audio_to_text import expand_sharded_filepaths
+from nemo.collections.asr.parts.utils.ipl_utils import (
+    create_final_cache_manifest,
+    expand_braces,
+    formulate_cache_manifest_names,
+    handle_multiple_tarr_filepaths,
+    process_manifest,
+    rm_punctuation,
+    sample_data,
+    write_cache_manifest,
+    write_tar_cache_manifest,
+)
+
+
+class IPLMixin:
+    """
+    Adds ability to do iterative pseudo-labeling.
+    To use ipl parameters should be given in the config as well as max_steps should be proovided in trainer.
+    For details, see: SlimIPL:(https://arxiv.org/pdf/2010.11524).
+
+    Parameters in config.
+    ipl:
+        m_epochs (int): Number of epochs to train model before first PL generation.
+        restore_pc (bool): Whether tp restore PC by comparing with already existing transcriptions if there are any. Defaults to `False`.
+        manifest_filepath (str): Path to the dataset manifest file.
+        tarred_audio_filepaths (str): Path to the tarred audio files.
+        is_tarred (bool): Flag indicating whether the dataset is tarred.
+        dataset_weights (float or list): What part of the dataset to use (applicable with non-tar datasets). Defaults to 1.
+        limit_train_batches (int): Limit_train_batches after PLs are added to train set (for lhotse only).
+        cache_manifest (str): Path for cache the manifest file.
+        dropout (float): Dropout rate used during training after PL generation.
+        n_l_epochs (int): Number of epochs to train the model with changed dropout before adding PLs to train set.
+        p_cache (float): Probability with which cache will be updated
+        cache_prefix (str): Prefix for cache files (optional for non-tar datasets).
+        batch_size (int): Batch size with which PLs will be generated
+
+    Call
+
+        * ``self.setup_ipl(_ipl_model_type)``
+          in the init method
+        * ``self.maybe_do_ipl()``
+          in the `on_train_epoch_end` method.
+    """
+
+    def setup_ipl(self, model_type: str):
+        """
+        Sets up IPL configuration for the model.
+
+        Args:
+            _ipl_model_type (str): The type of model being used. Takes values "hybrid" or "ctc".
+        """
+
+        ipl_config = self.cfg.get("ipl")
```

Comment:
Acronyms in a cfg is just asking for mistakes. Have the user write out `psuedo_label_cfg`.

Reply:
Done